system/security/LdapSecurity/ldapconnection.cpp

```diff
@@ -465,16 +464,17 @@ class CLdapConfig : implements ILdapConfig, public CInterface
 
             for(int retries = 0; retries <= LDAPSEC_MAX_RETRIES; retries++)
             {
+                getLdapHost(hostbuf);
                 rc = LdapUtils::getServerInfo(hostbuf.str(), sysUserDN.str(), m_sysuser_password.str(), m_protocol, port, m_cipherSuite, dcbuf, m_serverType, ldapDomain, m_timeout);
-                if(!LdapServerDown(rc) || retries >= LDAPSEC_MAX_RETRIES)
-                    break;
-                sleep(LDAPSEC_RETRY_WAIT);
-                if(retries < LDAPSEC_MAX_RETRIES)
+                if(rc == LDAP_TIMEOUT && retries <= LDAPSEC_MAX_RETRIES)
```

Comment:
This seems to deliberately be making the code as hard to follow as possible.  Why not

```
                if(rc != LDAP_TIMEOUT || retries >= LDAPSEC_MAX_RETRIES)
                    break;
                  sleep(LDAPSEC_RETRY_WAIT);
                  DBGLOG("LDAP AD Server %s temporarily unreachable for user %s on port %d, retrying...", hostbuf.str(), sysUserDN.str(), port);
```
Your code is also wrong - it will sleep even if it has matched the maximum retries.